vnext/Microsoft.ReactNative.Cxx/JSValue.h

```diff
@@ -0,0 +1,252 @@
+
+  JSValue() noexcept;
+  JSValue(std::nullptr_t) noexcept;
+  JSValue(JSValueObject &&value) noexcept;
+  JSValue(JSValueArray &&value) noexcept;
+  JSValue(std::string &&value) noexcept;
+  template <class TStringView, std::enable_if_t<std::is_convertible_v<TStringView, std::string_view>, int> = 1>
+  JSValue(TStringView value) noexcept;
+  template <class TBool, std::enable_if_t<std::is_same_v<TBool, bool>, int> = 1>
+  JSValue(TBool value) noexcept;
+  template <class TInt, std::enable_if_t<std::is_integral_v<TInt> && !std::is_same_v<TInt, bool>, int> = 1>
+  JSValue(TInt value) noexcept;
+  JSValue(double value) noexcept;
+
+  JSValue(const JSValue &other) = delete;
+  JSValue &operator=(const JSValue &other) = delete;
+
+  JSValue(JSValue &&other) noexcept;
+  JSValue &operator=(JSValue &&other) noexcept;
+
+  ~JSValue() noexcept;
+
+  JSValue Copy() const noexcept;
+  static JSValueObject CopyObject(const JSValueObject &other) noexcept;
+  static JSValueArray CopyArray(const JSValueArray &other) noexcept;
+
+  JSValueType Type() const noexcept;
+  bool IsNull() const noexcept;
+  const JSValueObject &Object() const noexcept;
+  const JSValueArray &Array() const noexcept;
+  const std::string &String() const noexcept;
+  bool Boolean() const noexcept;
+  int64_t Int64() const noexcept;
+  double Double() const noexcept;
+
+  JSValueObject TakeObject() noexcept;
+  JSValueArray TakeArray() noexcept;
+
+  size_t PropertyCount() const noexcept;
+  size_t ItemCount() const noexcept;
+
+  template <typename T>
+  T To() const noexcept;
+  template <class T>
+  static JSValue From(const T &value) noexcept;
+
+  const JSValue &GetObjectProperty(std::string_view propertyName) const noexcept;
+  const JSValue &GetArrayItem(JSValueArray::size_type index) const noexcept;
+  const JSValue &operator[](std::string_view propertyName) const noexcept;
+  const JSValue &operator[](JSValueArray::size_type index) const noexcept;
+
+  bool Equals(const JSValue &other) const noexcept;
+
+  static JSValue ReadFrom(IJSValueReader const &reader) noexcept;
+  static JSValueObject ReadObjectFrom(IJSValueReader const &reader) noexcept;
+  static JSValueArray ReadArrayFrom(IJSValueReader const &reader) noexcept;
+
+  void WriteTo(IJSValueWriter const &writer) const noexcept;
+  static void WriteObjectTo(IJSValueWriter const &writer, JSValueObjectView object) noexcept;
+  static void WriteArrayTo(IJSValueWriter const &writer, JSValueArrayView value) noexcept;
+
+ private:
+  bool ObjectEquals(const JSValueObject &other) const noexcept;
+  bool ArrayEquals(const JSValueArray &other) const noexcept;
+  static JSValueObject ReadObjectProperties(IJSValueReader const &reader) noexcept;
+  static JSValueArray ReadArrayItems(IJSValueReader const &reader) noexcept;
+
+ private: // Instance fields
+  JSValueType m_type;
+  union {
+    JSValueObject m_object;
+    JSValueArray m_array;
+    std::string m_string;
+    bool m_bool;
+    int64_t m_int64;
+    double m_double;
```

Comment:
Seems like we kind of just reimplemented std::variant. Using variant might make some other things easier, e.g. I'm pretty sure the default move constructor would work instead of switching on type like we do right now.

Reply:
It is more compact than use of std::variant. I do not see issues with implementing move constructor. The std::variant must have done it too, and we would not have any saving on code size. The std::variant seems to me as an overkill for the case when we can just use union.

---
In reply to: [356916675](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356916675) [](ancestors = 356916675)